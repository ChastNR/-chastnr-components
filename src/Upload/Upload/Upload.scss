.upl {
  &.ctrl__cntr {
    background-color: #fff;
    border: 1px dashed #428bf9;
    padding: 0;
    text-align: center;

    .ctrl__lb {
      width: 100%;
    }

    input[type='file'] {
      appearance: none;
      bottom: 0;
      cursor: pointer;
      height: 100%;
      left: 0;
      opacity: 0;
      position: absolute;
      right: 0;
      top: 0;
      width: 100%;
      z-index: 1;
    }

    &:hover {
      border-color: #428bf9;
      background-color: #fff;

      &.error {
        border-color: transparent;
        background-color: rgba(224, 31, 25, 0.16);
      }
    }

    &:focus-within {
      .ctrl__lb {
        color: rgba(0, 0, 0, 0.54);
        transform: unset;
      }

      &.error {
        background: #fff;
        border-color: #f52222;

        &.hasValue {
          .ctrl__lb {
            color: #e31c1c;
          }
        }
      }
    }
  }
}
