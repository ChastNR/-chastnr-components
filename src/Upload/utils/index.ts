export { createFileUploadError } from './createFileUploadError';
export { getFilenameAndExtension } from './getFilenameAndExtension';
export { formatBytes } from './formatBytes';
export { isFilesValid } from './isFilesValid';
