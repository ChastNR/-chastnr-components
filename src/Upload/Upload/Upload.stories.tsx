/* eslint-disable import/no-extraneous-dependencies */
import { useState } from 'react';
import { ComponentMeta } from '@storybook/react';
import { Upload } from './Upload';

const componentMeta: ComponentMeta<typeof Upload> = {
  title: 'Upload',
  component: Upload,
  parameters: {
    layout: 'fullscreen',
  },
};

export const DefaultUpload = () => {
  const [value, setValue] = useState<File[]>([]);

  return (
    <Upload
      accept=".jpg"
      files={value}
      info="Файлы pdf, jpeg, jpg, bmp, xlsx, docx не больше 10 Мб"
      label="Выберите файлы или перетяните их сюда"
      onChange={setValue}
    />
  );
};

export default componentMeta;
