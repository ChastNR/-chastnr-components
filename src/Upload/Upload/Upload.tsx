import { ChangeEvent, useState } from 'react';
import classNames from 'clsx';
import { Control } from '../../Control';
import { FilesInfo } from '../FilesInfo';
import { createFileUploadError } from '../utils';

import './Upload.scss';

interface IUploadProps {
  accept?: string;
  className?: string;
  disabled?: boolean;
  errorMessage?: React.ReactNode;
  files?: File[];
  info?: React.ReactNode;
  label: React.ReactNode;
  maxSizeKb?: number;
  multiple?: boolean;
  name?: string;
  onChange: (files: File[]) => void;
  requiredMessage?: React.ReactNode;
  startValidate?: boolean;
}

export const Upload: React.FC<IUploadProps> = ({
  accept,
  className,
  disabled,
  errorMessage,
  files = [],
  info,
  label,
  maxSizeKb,
  multiple,
  name,
  onChange,
  requiredMessage,
  startValidate = false,
}) => {
  const [validate, setValidate] = useState(false);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const file = e.currentTarget.files[0];

    if (!multiple) {
      onChange([file]);
    } else {
      const isTheSameFileExists = files.some((uploadedFile) => uploadedFile.name === file.name);
      const uploadedFiles = isTheSameFileExists ? files : [...files, file];
      onChange(uploadedFiles);
    }

    setValidate(true);
  };

  const handleDelete = (fileName: string) => {
    const uploadedFiles = files.filter((uploadedFile) => uploadedFile.name !== fileName);
    onChange(uploadedFiles);
  };

  const error = createFileUploadError({
    accept,
    validate,
    startValidate,
    errorMessage,
    requiredMessage,
    files,
    maxSizeKb,
  });

  const controlClassName = classNames('upl', className);

  return (
    <Control
      className={controlClassName}
      disabled={disabled}
      error={error}
      info={<FilesInfo accept={accept} files={files} info={info} onDeleteClick={handleDelete} />}
      label={label}
      name={name}
      required={!!requiredMessage}
    >
      <input
        accept={accept}
        disabled={disabled}
        multiple={multiple}
        name={name}
        onChange={handleChange}
        required={!!requiredMessage}
        type="file"
      />
    </Control>
  );
};
