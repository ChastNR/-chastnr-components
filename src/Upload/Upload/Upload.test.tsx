import { fireEvent, render } from '@testing-library/react';
import { Upload } from './Upload';

describe('Upload', () => {
  const testFile = new File(['test'], 'test.txt', {
    type: 'text/plain',
  });

  it('should render properly', () => {
    const mockOnChange = jest.fn();

    const upload = render(
      <Upload
        accept=".txt"
        className="testClassName"
        info="testInfo"
        errorMessage="error"
        files={[testFile]}
        label="Upload label"
        maxSizeKb={1000}
        multiple
        name="Hi"
        onChange={mockOnChange}
        requiredMessage="Files required"
      />
    );

    expect(upload.container.firstChild).toMatchSnapshot();
  });

  it('should add and delete file when multiple is truthy', () => {
    const newFile = new File(['test1'], 'test1.txt', {
      type: 'text/plain',
    });

    let files: File[] = [testFile];

    const mockOnChange = jest.fn((value: File[]) => (files = value));

    const upload = render(
      <Upload
        accept=".txt"
        className="testClassName"
        info="testInfo"
        errorMessage="error"
        files={files}
        label="Upload label"
        maxSizeKb={1000}
        multiple
        name="Hi"
        onChange={mockOnChange}
        requiredMessage="Files required"
      />
    );

    const inputElement = upload.container.getElementsByTagName('input')[0];
    const buttonElement = upload.container.getElementsByTagName('button')[0];

    fireEvent.change(inputElement, { target: { files: [newFile] } });
    fireEvent.click(buttonElement);

    expect(mockOnChange).toBeCalledWith([testFile, newFile]);
    expect(mockOnChange).toBeCalledTimes(2);
  });

  it('should add file when multiple is falsy', () => {
    const newFile = new File(['test1'], 'test1.txt', {
      type: 'text/plain',
    });

    let files: File[] = [testFile];

    const mockOnChange = jest.fn((value: File[]) => (files = value));

    const upload = render(
      <Upload
        accept=".txt"
        className="testClassName"
        info="testInfo"
        errorMessage="error"
        files={files}
        label="Upload label"
        maxSizeKb={1000}
        name="Hi"
        onChange={mockOnChange}
        requiredMessage="Files required"
      />
    );

    const inputElement = upload.container.getElementsByTagName('input')[0];

    fireEvent.change(inputElement, { target: { files: [newFile] } });

    expect(mockOnChange).toBeCalledWith([newFile]);
  });

  it('should return current files when same file is exists', () => {
    let files: File[] = [testFile];

    const mockOnChange = jest.fn((value: File[]) => (files = value));

    const upload = render(
      <Upload
        accept=".txt"
        className="testClassName"
        info="testInfo"
        errorMessage="error"
        files={files}
        label="Upload label"
        maxSizeKb={1000}
        name="Hi"
        onChange={mockOnChange}
        requiredMessage="Files required"
      />
    );

    const inputElement = upload.container.getElementsByTagName('input')[0];

    fireEvent.change(inputElement, { target: { files: [testFile] } });

    expect(mockOnChange).toBeCalledWith([testFile]);
  });
});
