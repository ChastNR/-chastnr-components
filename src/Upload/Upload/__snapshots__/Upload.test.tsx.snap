// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload should render properly 1`] = `
<div
  class="ctrl"
>
  <div
    class="ctrl__cntr upl testClassName required"
    role="textbox"
    tabindex="0"
  >
    <label
      class="ctrl__lb"
      for="Hi"
    >
      Upload label
    </label>
    <input
      accept=".txt"
      multiple=""
      name="Hi"
      required=""
      type="file"
    />
  </div>
  <div>
    <ul
      class="upl__fls"
    >
      <li
        class="upl__fl"
      >
        <div
          class="upl__fl__cntr"
        >
          <div
            class="upl__fl__cntr__vldt valid"
          >
            <span
              class="rnd__tick"
            >
              <span>
                content
              </span>
            </span>
          </div>
          <div
            class="upl__fl__cntr__data"
          >
            <span
              class="upl__fl__cntr__name"
            >
              test
            </span>
            <span>
              .
              txt
            </span>
            <span
              class="upl__fl__cntr__size"
            >
              0.00 KB
            </span>
          </div>
          <button
            class="upl__fl__cntr__dlt__btn"
            type="button"
          >
            <span
              class="close"
            >
              <span>
                content
              </span>
            </span>
          </button>
        </div>
      </li>
    </ul>
    <div
      class="upl__info"
    >
      testInfo
    </div>
  </div>
  <div
    class="er__cntr"
    style="height: 0px;"
  >
    <div />
  </div>
</div>
`;
